PidController python binding seems to have memory corruption

Add the following line:
```
        self.assertEqual(controller.num_input_ports(), 2)
```
to
https://github.com/RobotLocomotion/drake/blob/master/bindings/pydrake/systems/test/controllers_test.py#L221

results in
```
test_pid_controller
    self.assertEqual(controller.num_input_ports(), 2)
AssertionError: -751256949 != 2
```

--- a/bindings/pydrake/systems/controllers_py.cc
+++ b/bindings/pydrake/systems/controllers_py.cc
@@ -74,6 +74,8 @@ PYBIND11_MODULE(controllers, m) {
           doc.InverseDynamics.InverseDynamicsMode.kGravityCompensation.doc)
       .export_values();
 
+  // TODO(eric.cousineau): Expose multiple inheritance from
+  // StateFeedbackControllerInterface once #9243 is resolved.
   py::class_<InverseDynamicsController<double>, Diagram<double>>(
       m, "InverseDynamicsController", doc.InverseDynamicsController.doc)
       .def(py::init<const MultibodyPlant<double>&, const VectorX<double>&,
@@ -144,6 +146,8 @@ PYBIND11_MODULE(controllers, m) {
           py_reference_internal,
           doc.PidControlledSystem.get_state_output_port.doc);
 
+  // TODO(eric.cousineau): Expose multiple inheritance from
+  // StateFeedbackControllerInterface once #9243 is resolved.
   py::class_<PidController<double>, LeafSystem<double>>(
       m, "PidController", doc.PidController.doc)
       .def(py::init<const VectorX<double>&, const VectorX<double>&,

--- a/bindings/pydrake/systems/test/controllers_test.py
+++ b/bindings/pydrake/systems/test/controllers_test.py
@@ -191,40 +191,47 @@ class TestControllers(unittest.TestCase):
                                     tau_id))
 
     def test_pid_controlled_system(self):
-        controller = PidControlledSystem(plant=PendulumPlant(), kp=1., ki=0.,
-                                         kd=2., state_output_port_index=0)
-        controller = PidControlledSystem(plant=PendulumPlant(), kp=[0], ki=[1],
-                                         kd=[2], state_output_port_index=0)
-        controller = PidControlledSystem(plant=PendulumPlant(),
-                                         feedback_selector=np.eye(2), kp=1.,
-                                         ki=0., kd=2.,
-                                         state_output_port_index=0)
-        controller = PidControlledSystem(plant=PendulumPlant(),
-                                         feedback_selector=np.eye(2),
-                                         kp=[0], ki=[1], kd=[2],
-                                         state_output_port_index=0)
+        controllers = [
+            PidControlledSystem(plant=PendulumPlant(), kp=1., ki=0.,
+                                kd=2., state_output_port_index=0),
+            PidControlledSystem(plant=PendulumPlant(), kp=[0], ki=[1],
+                                kd=[2], state_output_port_index=0),
+            PidControlledSystem(plant=PendulumPlant(),
+                                feedback_selector=np.eye(2), kp=1.,
+                                ki=0., kd=2.,
+                                state_output_port_index=0),
+            PidControlledSystem(plant=PendulumPlant(),
+                                feedback_selector=np.eye(2),
+                                kp=[0], ki=[1], kd=[2],
+                                state_output_port_index=0),
+        ]
 
-        self.assertIsNotNone(controller.get_control_input_port())
-        self.assertIsNotNone(controller.get_state_input_port())
-        self.assertIsNotNone(controller.get_state_output_port())
+        for controller in controllers:
+            self.assertIsNotNone(controller.get_control_input_port())
+            self.assertIsNotNone(controller.get_state_input_port())
+            self.assertIsNotNone(controller.get_state_output_port())
 
     def test_pid_controller(self):
-        controller = PidController(kp=np.ones(3), ki=np.zeros(3),
-                                   kd=[1, 2, 3])
-        controller = PidController(state_projection=np.ones((6, 4)),
-                                   kp=np.ones(3), ki=np.zeros(3),
-                                   kd=[1, 2, 3])
-        controller = PidController(state_projection=np.ones((6, 4)),
-                                   output_projection=np.ones((4, 3)),
-                                   kp=np.ones(3), ki=np.zeros(3),
-                                   kd=[1, 2, 3])
+        controllers = [
+            PidController(kp=np.ones(3), ki=np.zeros(3),
+                          kd=[1, 2, 3]),
+            PidController(state_projection=np.ones((6, 4)),
+                          kp=np.ones(3), ki=np.zeros(3),
+                          kd=[1, 2, 3]),
+            PidController(state_projection=np.ones((6, 4)),
+                          output_projection=np.ones((4, 3)),
+                          kp=np.ones(3), ki=np.zeros(3),
+                          kd=[1, 2, 3]),
+        ]
 
-        self.assertEqual(len(controller.get_Kp_vector()), 3)
-        self.assertEqual(len(controller.get_Ki_vector()), 3)
-        self.assertEqual(len(controller.get_Kd_vector()), 3)
-        self.assertIsNotNone(controller.get_input_port_estimated_state())
-        self.assertIsNotNone(controller.get_input_port_desired_state())
-        self.assertIsNotNone(controller.get_output_port_control())
+        for controller in controllers:
+            self.assertEqual(controller.num_input_ports(), 2)
+            self.assertEqual(len(controller.get_Kp_vector()), 3)
+            self.assertEqual(len(controller.get_Ki_vector()), 3)
+            self.assertEqual(len(controller.get_Kd_vector()), 3)
+            self.assertIsNotNone(controller.get_input_port_estimated_state())
+            self.assertIsNotNone(controller.get_input_port_desired_state())
+            self.assertIsNotNone(controller.get_output_port_control())
 
     def test_linear_quadratic_regulator(self):
         A = np.array([[0, 1], [0, 0]])

--- a/examples/kuka_iiwa_arm/kuka_torque_controller.h
+++ b/examples/kuka_iiwa_arm/kuka_torque_controller.h
@@ -12,6 +12,7 @@ namespace drake {
 namespace examples {
 namespace kuka_iiwa_arm {
 
+// N.B. Inheritance order must remain fixed for pydrake (#9243).
 /**
  * Controller that take emulates the kuka_iiwa_arm when operated in torque
  * control mode. The controller specifies a stiffness and damping ratio at each

--- a/examples/qp_inverse_dynamics/manipulator_joint_space_controller.h
+++ b/examples/qp_inverse_dynamics/manipulator_joint_space_controller.h
@@ -13,6 +13,7 @@ namespace drake {
 namespace examples {
 namespace qp_inverse_dynamics {
 
+// N.B. Inheritance order must remain fixed for pydrake (#9243).
 /**
  * This class implements an example of a general purpose controller for a
  * fixed based manipulator based on qp inverse dynamics. It contains two main
@@ -32,8 +33,8 @@ namespace qp_inverse_dynamics {
  * implemented by different plan eval modules.
  */
 class ManipulatorJointSpaceController
-    : public systems::controllers::StateFeedbackControllerInterface<double>,
-      public systems::Diagram<double> {
+    : public systems::Diagram<double>,
+      public systems::controllers::StateFeedbackControllerInterface<double> {
  public:
   /**
    * Constructs a inverse dynamics controller for a fixed base manipulator that

--- a/manipulation/schunk_wsg/schunk_wsg_plain_controller.h
+++ b/manipulation/schunk_wsg/schunk_wsg_plain_controller.h
@@ -10,6 +10,7 @@ namespace schunk_wsg {
 
 enum class ControlMode { kPosition = 0, kForce = 1 };
 
+// N.B. Inheritance order must remain fixed for pydrake (#9243).
 /** This class implements a controller for a Schunk WSG gripper as a
  * `systems::Diagram`. The composition of this diagram is determined by the
  * control mode specified for the controller, which can be either

--- a/systems/controllers/inverse_dynamics_controller.h
+++ b/systems/controllers/inverse_dynamics_controller.h
@@ -16,6 +16,7 @@ namespace drake {
 namespace systems {
 namespace controllers {
 
+// N.B. Inheritance order must remain fixed for pydrake (#9243).
 /**
  * A state feedback controller that uses a PidController to generate desired
  * accelerations, which are then converted into torques using InverseDynamics.

--- a/systems/controllers/pid_controller.h
+++ b/systems/controllers/pid_controller.h
@@ -14,6 +14,7 @@ namespace controllers {
 // TODO(siyuanfeng): Lift the assumption that q and v have the same dimension.
 // TODO(siyuanfeng): Generalize "q_d - q", e.g. for rotation.
 
+// N.B. Inheritance order must remain fixed for pydrake (#9243).
 /**
  * Implements the PID controller. Given estimated state `x_in = (q_in, v_in)`,
  * the controlled state `x_c = (q_c, v_c)` is computed by `x_c = P_x * x_in`,
@@ -43,8 +44,8 @@ namespace controllers {
  * @ingroup control_systems
  */
 template <typename T>
-class PidController : public StateFeedbackControllerInterface<T>,
-                      public LeafSystem<T> {
+class PidController : public LeafSystem<T>,
+                      public StateFeedbackControllerInterface<T> {
  public:
   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(PidController)
 

